patches/net/minecraft/client/multiplayer/ClientConfigurationPacketListenerImpl.java.patch

```diff
@@ -4,67 +4,95 @@
-+    private boolean isVanillaConnection = true;
++    private net.neoforged.neoforge.network.connection.ConnectionType connectionType = net.neoforged.neoforge.network.connection.ConnectionType.VANILLA;
++    private boolean initializedConnection = false;
 +    private java.util.Map<net.minecraft.resources.ResourceLocation, net.minecraft.network.chat.Component> failureReasons = new java.util.HashMap<>();
  
      public ClientConfigurationPacketListenerImpl(Minecraft p_295262_, Connection p_296339_, CommonListenerCookie p_294706_) {
          super(p_295262_, p_296339_, p_294706_);
-@@ -45,7 +_,30 @@
+@@ -45,7 +_,41 @@
      }
  
      @Override
 -    protected void handleCustomPayload(CustomPacketPayload p_295411_) {
 +    protected void handleCustomPayload(net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket p_295727_, CustomPacketPayload p_295411_) {
++        if (p_295411_ instanceof net.neoforged.neoforge.network.payload.MinecraftRegisterPayload minecraftRegisterPayload) {
++            this.connectionType = this.connectionType.withMinecraftRegisterPayload();
++            net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().onMinecraftRegister(this, minecraftRegisterPayload.newChannels());
++            return;
++        }
++        if (p_295411_ instanceof net.neoforged.neoforge.network.payload.MinecraftUnregisterPayload minecraftUnregisterPayload) {
++            this.connectionType = this.connectionType.withMinecraftRegisterPayload();
++            net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().onMinecraftUnregister(this, minecraftUnregisterPayload.forgottenChannels());
++            return;
++        }
 +        if (p_295411_ instanceof net.neoforged.neoforge.network.payload.ModdedNetworkQueryPayload) {
-+            this.isModdedConnection = true;
++            this.connectionType = this.connectionType.withNeoForgeQueryPayload();
 +            net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().onNetworkQuery(this);
 +            return;
 +        }
 +        if (p_295411_ instanceof net.neoforged.neoforge.network.payload.ModdedNetworkPayload moddedNetworkPayload) {
++            this.initializedConnection = true;
 +            net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().onModdedNetworkConnectionEstablished(this, moddedNetworkPayload.configuration(), moddedNetworkPayload.play());
 +            return;
 +        }
 +        if (p_295411_ instanceof net.neoforged.neoforge.network.payload.ModdedNetworkSetupFailedPayload setupFailedPayload) {
 +            failureReasons = setupFailedPayload.failureReasons();
 +        }
-+        if (!this.isModdedConnection && p_295411_ instanceof net.minecraft.network.protocol.common.custom.BrandPayload) {
-+            this.isVanillaConnection = true;
++        if (!this.connectionType.isNeoForge() && p_295411_ instanceof net.minecraft.network.protocol.common.custom.BrandPayload) {
```

Comment:
I don't like relying on Brand. We should just check the connection type when we receive Ping.

Reply:
Sadly it is what is.
We discussed this very early on even before the original port, but during testing and processing this was the best we could come up with, we tried the approach on just relying on the ping packet on the client side, but it simply was not stable enough.

Reply:
Fair enough